src/chai/ArrayManager.cpp

```diff
@@ -137,16 +149,25 @@ void ArrayManager::move(PointerRecord* record, ExecutionSpace space)
     return;
   }
 
+  if (space == record->m_last_space) {
+    return;
+  }
+
 #if defined(CHAI_ENABLE_UM)
   if (record->m_last_space == UM) {
     return;
   }
 #endif
 
-  if (space == record->m_last_space) {
+#if defined(CHAI_ENABLE_PINNED)
+  if (record->m_last_space == PINNED) {
+    if (space == CPU && m_need_sync_for_pinned) {
+      m_need_sync_for_pinned = false;
+      cudaDeviceSynchronize();
+    }
     return;
```

Comment:
I'd add some documentation to the early returns in this explaining why no move is required  (I had to look at the full file context to figure out this was where we were circumventing the copy).

Reply:
Will do, thanks for the review.